torch/utils/data/standard_pipes.ipynb

```diff
@@ -0,0 +1,1006 @@
+    "\n",
+    "Classic batching produce partial batches by default\n"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[0, 1, 2]\n[3, 4, 5]\n[6, 7, 8]\n[9]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(10).batch(3)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "To drop incomplete batches add `drop_last` argument"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[0, 1, 2]\n[3, 4, 5]\n[6, 7, 8]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(10).batch(3, drop_last=True)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "Without `batch_level` override sequential `batch` calls do rebatching."
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[0, 1]\n[2, 3]\n[4, 5]\n[6, 7]\n[8, 9]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(10).batch(3).batch(2)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "Setting `batch_level=True` allows to take input lists as singular objects and nest them"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[[0, 1, 2], [3, 4, 5]]\n[[6, 7, 8], [9]]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(10).batch(3).batch(2, batch_level=True)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "## Map\n",
+    "\n",
+    "Function: `map`\n",
+    "\n",
+    "Description: \n",
+    "\n",
+    "Alternatives:\n",
+    "\n",
+    "Arguments:\n",
+    "  - `batch_level: bool = False` whether elements from the source DataPipe counted as independant objects or we process mini-batched elements as whole.\n",
+    " \n",
+    "Example:"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "0\n2\n4\n6\n8\n10\n12\n14\n16\n18\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(10).map(lambda x: x * 2)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "Classic `map` applies function to every element inside mini-batches by default\n"
```

Comment:
so, if i do `[1, 2.. 100].batch(n1).batch(n2, batch_level=True).batch(n3, batch_level=True).map(x -> x *2)`, the final `map` will "penetrates" two levels of batch and still applies to each number?

Reply:
No, it only penetrates one level.